Framework/PythonInterface/plugins/algorithms/FindUBFromScatteringPlane.py

```diff
@@ -0,0 +1,135 @@
+        )
+
+        self.declareProperty(
+            name="alpha", direction=Direction.Input, validator=angle_limit, defaultValue=90.0, doc="Lattice Parameter alpha"
+        )
+        self.declareProperty(name="beta", direction=Direction.Input, validator=angle_limit, defaultValue=90.0, doc="Lattice Parameter beta")
+        self.declareProperty(
+            name="gamma", direction=Direction.Input, validator=angle_limit, defaultValue=90.0, doc="Lattice Parameter gamma"
+        )
+
+    def exec_child_alg(self, alg_name: str, **kwargs):
+        alg = self.createChildAlgorithm(alg_name, enableLogging=False)
+        alg.setAlwaysStoreInADS(False)
+        alg.initialize()
+        alg.setProperties(kwargs)
+        alg.execute()
+        out_props = tuple(alg.getProperty(prop).value for prop in alg.outputProperties())
+        return out_props[0] if len(out_props) == 1 else out_props
+
+    # Can change these to be 2 arrays of abc, and alpha,beta,gamma if that is preferred
+    def PyExec(self):
+        # getting lattice parameters
+        a = self.getProperty("a").value
+        b = self.getProperty("b").value
+        c = self.getProperty("c").value
+
+        alpha = self.getProperty("alpha").value
+        beta = self.getProperty("beta").value
+        gamma = self.getProperty("gamma").value
+
+        # getting vectors to define the plane
+        vector_1 = np.array(self.getProperty("Vector1").value)
+        vector_1 = V3D(vector_1[0], vector_1[1], vector_1[2])
+        vector_2 = np.array(self.getProperty("Vector2").value)
+        vector_2 = V3D(vector_2[0], vector_2[1], vector_2[2])
+
+        # Find a UB with the correct scattering plane (even if `u` and `v` passed to `SetUB` are not orthogonal)
+
+        peak_table = self.getProperty("PeaksWorkspace").value
+        self.exec_child_alg("SetUB", Workspace=peak_table, u=vector_1, v=vector_2, a=a, b=b, c=c, alpha=alpha, beta=beta, gamma=gamma)
+        lattice = peak_table.sample().getOrientedLattice()
+        ub_arb = lattice.getUB()
+
+        vertical_dir = peak_table.getInstrument().getReferenceFrame().vecPointingUp()
+        vertical_dir = vertical_dir / np.linalg.norm(np.array(vertical_dir))
+        vertical_dir = V3D(vertical_dir[0], vertical_dir[1], vertical_dir[2])
+
+        peak_ip = peak_table.getPeak(0)
+        hkl_peak = np.array(peak_ip.getHKL())
+        hkl_peak = hkl_peak.reshape(-1, 1)
+        qsample_calc = np.array(np.pi * 2 * (ub_arb @ hkl_peak))
+        qsample_calc = V3D(float(qsample_calc[0]), float(qsample_calc[1]), float(qsample_calc[2]))
+        qsample_obs = np.array(peak_ip.getQSampleFrame())
+        qsample_obs = V3D(float(qsample_obs[0]), float(qsample_obs[1]), float(qsample_obs[2]))
+
+        if not np.degrees(qsample_calc.angle(vertical_dir)) < 5:
+            logger.warning("given peak hkl does not lie in the plane")
+
+        def rotation(theta):
```

Comment:
Don't think this function is used anywhere outside `objective` now, so as it's small could just go inside `objective`